docs/internal/expressions.md

```diff
@@ -0,0 +1,597 @@
+
+// Types can be alias to sub-types
+interface Foo {}
+class Bar : Foo {}
+
+let d : Bar & mut = Foo;
+let e : Foo & mut = d;
+
+let eq = (d is e); // TRUE, `e` is an alias to `d` of a compatible type
+
+// Singleton types are always equal to themselves but not other singleton types
+class None {}
+class All {}
+
+let always_eq = (None is None); // TRUE
+let never_eq = (None is All); // FALSE
+```
+
+Because singleton types (other than numeric types) have no machine representation, they always compare equal on the same types, which means all variables of the same singleton type are aliases to each other.
+
+Note that, even though literals have a machine representation for their values, the addresses that they are stored for `a` and `c` are different, and therefore they are not the _same variable_, even though they have the same value.
+
+### Structural equality
+
+Structural equality, compared with the method `==` on appropriate types, is a comparison of the values in the object that the variable points to, including numeric literal values.
+It is the responsibility of each class to implement their own structural equality method.
+
+The standard library implements them for all classes that can be structurally compared.
+User objects can be compared structurally, simply by implementing the `==` method.
+Similar operators (`!=`, `<`, `>`, etc) are also implemented in the standard library and can be used by any user class in the same way.
+
+Example:
+```ts
+// Some variable
+var a : U32 & mut = 42;
+// A new variable with different value
+let b = a + 1;
+// A new variable with the same value (but different address)
+var c : U32 & mut = 42;
+
+let ne = (a == b); // FALSE, 42 != 43
+let eq = (a == c); // TRUE, both as 42
+```
+
+_Note: There is nothing special about structural comparison methods.
+ They're methods like any other that just happens to "implement" comparisons.
+ You can do whatever you want with your own types for `==`._
+
+## Syntax Sugar
+
+In Verona, every operation is a method call, including lambda execution, object creation, destruction, etc.
+To create syntactically correct code, one must be explicit about everything.
+
+Example:
+```ts
+class MyType {
+  create() : MyType;
+  apply();
+}
+// Creates a new object and calls it
+let a : MyType & mut = MyType::create();
+a.apply();
+```
+
+To make the code above more readable, Verona introduces _syntax sugar_, which allows you to write more concise code meaning the same thing.
+
+The two main sugars are:
+* `create`, for constructing objects, allowing you to call just the class name instead of the whole call syntax.
+* `apply`, for calling objects as functions, including lambdas, without the need to call it directly.
+
+The code above would look like:
+```ts
+let a : MyType & mut = MyType;
+a();
+```
+_Note: We also want the `update` sugar, but it may need a `Ref[T]` type to do `obj.update(field) = val` where `update()` returns `Ref[T]`._
```

Comment:
I thought the point of `update` sugar was to avoid the need for refs.

Reply:
this is a snippet from one of @sylvanc meetings. I'll let him answer that.